### PR TITLE
Adding various docs to clarify structs / traits

### PR DESCRIPTION
**Pull request by jnwng**
Intended to be viewed via `docs.rs`. Locally, run `cargo doc --open` to preview

### CODE REVIEW

File: contracts/programs/gummyroll/src/state/mod.rs [hunk 1]
```diff
@@ -65,10 +72,23 @@ impl<const MAX_DEPTH: usize> From<(Box<ChangeLog<MAX_DEPTH>>, Pubkey, u64)>
...
     pub max_depth: u32,
+
+    /// Authority that validates the content of the trees.
+    /// Typically a program, e.g., the Bubblegum contract validates that leaves are valid NFTs.
     pub authority: Pubkey,
+
+    /// Authority that is responsible for signing for new additions to the tree.
     pub append_authority: Pubkey,
+
+    /// TODO(jon): Not sure what this is!
```
Comment (jarry-xiao) [inline contracts/programs/gummyroll/src/state/mod.rs:91]: The idea is that we can use this information in the indexer to know how far back to look if we need to rebuild the tree from scratch

File: contracts/programs/gummyroll/src/state/mod.rs [hunk 2]
```diff
@@ -65,10 +72,23 @@ impl<const MAX_DEPTH: usize> From<(Box<ChangeLog<MAX_DEPTH>>, Pubkey, u64)>
...
+    /// Depth of the Merkle tree to store.
+    /// Valid options are any integer between 14-30
+    /// Tree capacity can be calculated as power(2, max_depth)
     pub max_depth: u32,
+
+    /// Authority that validates the content of the trees.
+    /// Typically a program, e.g., the Bubblegum contract validates that leaves are valid NFTs.
     pub authority: Pubkey,
+
+    /// Authority that is responsible for signing for new additions to the tree.
```
Comment (jarry-xiao) [inline contracts/programs/gummyroll/src/state/mod.rs:88]: This paradigm will likely be removed to just a single authority. Need to figure out exactly how that will look

**Review by ngundotra — APPROVED** [review]
Looks great! Thanks for setting this up.

It would be great to explicitly list out the valid `(max_depth, max_buffer_size)` tuples in a single place. Also would be great to link some tree visualization from Jarry's paper.

File: lib/concurrent-merkle-tree/src/error.rs [hunk 3]
```diff
@@ -2,20 +2,35 @@ use thiserror::Error;
...
+
+    /// Invalid number of bytes passed for node (expected 32 bytes)
     #[error("Invalid number of bytes passed for node (expected 32 bytes)")]
     InvalidNodeByteLength,
+
+    /// Root not found in changelog buffer
     #[error("Root not found in changelog buffer")]
     RootNotFound,
+
+    /// Valid proof was passed to a leaf, but it's value has changed since the proof was issued
```
[the remark this answers is not in the document]
Comment (jnwng) [reply lib/concurrent-merkle-tree/src/error.rs]: i mostly copied the errors if i thought they were self-explanatory, but could use some help adding more color to the errors that have more specific context

File: contracts/programs/gummyroll/src/state/mod.rs [hunk 4]
```diff
@@ -65,10 +72,23 @@ impl<const MAX_DEPTH: usize> From<(Box<ChangeLog<MAX_DEPTH>>, Pubkey, u64)>
 #[derive(BorshDeserialize, BorshSerialize)]
 #[repr(C)]
 pub struct MerkleRollHeader {
+    /// Buffer of changelogs stored on-chain. Must be a power of 2;
+    /// Valid options limited to: `(8, 64, 256, 512, 1024, 2048)`
     pub max_buffer_size: u32,
+
+    /// Depth of the Merkle tree to store.
+    /// Valid options are any integer between 14-30
```
[the remark this answers is not in the document]
Comment (jnwng) [reply contracts/programs/gummyroll/src/state/mod.rs]: image

File: contracts/programs/gummyroll/src/error.rs [hunk 5]
```diff
@@ -1,16 +1,29 @@
...
     #[msg("Concurrent merkle tree error")]
     ConcurrentMerkleTreeError,
+
+    /// An issue was detected with loading the provided account data for this Gummyroll tree.
     #[msg("Issue zero copying concurrent merkle tree data")]
     ZeroCopyError,
+
+    /// The Gummyroll tree supports two configuration parameters: `max_depth` and `max_buffer_size`.
+    /// - `max_buffer_size` must be a power of 2; valid values are limited to `(8, 64, 256, 512, 1024, 2048)`
+    /// - `max_depth` can be any integer between 14 and 30.
```
[the remark this answers is not in the document]
Comment (jnwng) [reply contracts/programs/gummyroll/src/error.rs]: done!

**Comment by ngundotra** [comment]
Looks great!

**Pull request merged**